crates/assists/src/utils/insert_use.rs

```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
+                let to_delete: SyntaxElement = existing_use.syntax().clone().into();
+                let to_delete = to_delete.clone()..=to_delete;
+                let to_insert = iter::once(merged.syntax().clone().into());
+                return algo::replace_children(&where_, to_delete, to_insert);
+            }
+        }
+    }
+
+    // either we weren't allowed to merge or there is no import that fits the merge conditions
+    // so look for the place we have to insert to
+    let (insert_position, add_blank) = find_insert_position(&where_, path);
+
+    let to_insert: Vec<SyntaxElement> = {
+        let mut buf = Vec::new();
+
+        if add_blank == AddBlankLine::Before {
+            buf.push(make::tokens::single_newline().into());
+        }
+
+        buf.push(use_item.syntax().clone().into());
+
+        if add_blank == AddBlankLine::After {
+            buf.push(make::tokens::single_newline().into());
+        } else if add_blank == AddBlankLine::AfterTwice {
+            buf.push(make::tokens::single_newline().into());
+            buf.push(make::tokens::single_newline().into());
+        }
+
+        buf
+    };
 
-    if let Some(container) = container {
-        let syntax = container.either(|l| l.syntax().clone(), |r| r.syntax().clone());
-        let action = best_action_for_target(syntax, position.clone(), &target);
-        make_assist(&action, &target, builder);
+    algo::insert_children(&where_, insert_position, to_insert)
+}
+
+fn try_merge_imports(
+    old: &ast::Use,
+    new: &ast::Use,
+    merge_behaviour: MergeBehaviour,
+) -> Option<ast::Use> {
+    // dont merge into re-exports
+    if old.visibility().map(|vis| vis.pub_token()).is_some() {
+        return None;
+    }
+    let old_tree = old.use_tree()?;
+    let new_tree = new.use_tree()?;
+    let merged = try_merge_trees(&old_tree, &new_tree, merge_behaviour)?;
+    Some(old.with_use_tree(merged))
+}
+
+/// Simple function that checks if a UseTreeList is deeper than one level
+fn use_tree_list_is_nested(tl: &ast::UseTreeList) -> bool {
+    tl.use_trees().any(|use_tree| {
+        use_tree.use_tree_list().is_some() || use_tree.path().and_then(|p| p.qualifier()).is_some()
+    })
+}
+
+pub fn try_merge_trees(
+    old: &ast::UseTree,
+    new: &ast::UseTree,
+    merge_behaviour: MergeBehaviour,
+) -> Option<ast::UseTree> {
+    let lhs_path = old.path()?;
+    let rhs_path = new.path()?;
+
+    let (lhs_prefix, rhs_prefix) = common_prefix(&lhs_path, &rhs_path)?;
+    let lhs = old.split_prefix(&lhs_prefix);
+    let rhs = new.split_prefix(&rhs_prefix);
+    let lhs_tl = lhs.use_tree_list()?;
+    let rhs_tl = rhs.use_tree_list()?;
+
+    // if we are only allowed to merge the last level check if the split off paths are only one level deep
+    if merge_behaviour == MergeBehaviour::Last && use_tree_list_is_nested(&lhs_tl)
+        || use_tree_list_is_nested(&rhs_tl)
+    {
+        return None;
```

Comment:
Such special cases could make use of marks: https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/test_utils/src/mark.rs

Reply:
Ah that seems nice, I actually haven't written a test for this case I think haha